src/search_cuda.cpp

```diff
@@ -62,27 +57,12 @@ BeamSearch_Cuda::BeamSearch_Cuda(const GeneratorParams& params)
 BeamSearch_Cuda::~BeamSearch_Cuda() = default;
 
 void Search_Cuda::SetLogits(RoamingArray<float> logits_unk) {
-  gpu_span<float> logits = logits_unk;
-  // Logits has shape (batch_size, input_length, vocab_size),
-  // where input_length equals to parameters_->sequence_length for first subgraph call, and 1 for the remaining calls.
-  // RyanHill: Does it really? The output of gpt2 is always a input_length of 1, regardless of input sequence length
-
+  next_token_scores_ = logits_unk.GetGPU();
   auto batch_beam_size = params_.BatchBeamSize();
-  auto input_length = logits.size() / (batch_beam_size * params_.vocab_size);
-  assert(logits.size() % (batch_beam_size * params_.vocab_size) == 0);  // Should divide evenly
 
-  // TODO: if input_length==1, use token scores directly
-
-  // Get logits for the last token:
-  //    next_token_logits = logits[:, -1, :], and the result shape is (batch_size, vocab_size)
-  // When input_length == 1, use logits directly in SoftmaxCPU below so it only need for input_length > 1.
-  const float* current_logits = logits.data() + (input_length - 1) * params_.vocab_size;
+  // Apply log_softmax to the input logits
   for (int i = 0; i < batch_beam_size; i++) {
-    std::span<const float> source(current_logits, params_.vocab_size);
     std::span<float> target = next_token_scores_.subspan(i * params_.vocab_size, params_.vocab_size);
-    CudaCheck() == cudaMemcpyAsync(target.data(), source.data(), source.size_bytes(), cudaMemcpyDeviceToDevice, params_.cuda_stream);
-    current_logits += input_length * params_.vocab_size;
-
     cuda::Launch_log_softmax(target.data(), static_cast<int>(target.size()), params_.cuda_stream);
   }
```

Comment:
The implementation of Launch_log_softmax is quite naive. The performance will be bad.

Reply:
The Launch_log_softmax does nothing:
[https://github.com/microsoft/onnxruntime-genai/blob/dbd10985e109cbf76f9f67f4e939c26eacc88288/src/search_cuda.cu#L41-L54](https://github.com/microsoft/onnxruntime-genai/blob/dbd10985e109cbf76f9f67f4e939c26eacc88288/src/search_cuda.cu#L41-L54)

Reply:
We don't need it on GPU as the Sampling handles it.

Reply:
[https://github.com/microsoft/onnxruntime-genai/blob/2addfc6f4e6c307a9329d548078003c2dabf15ff/src/cuda_sampling.cu#L582](https://github.com/microsoft/onnxruntime-genai/blob/2addfc6f4e6c307a9329d548078003c2dabf15ff/src/cuda_sampling.cu#L582)

Reply:
That's right, I'll remove these parts as they're not doing any work.